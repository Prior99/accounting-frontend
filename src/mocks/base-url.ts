declare var global: any;
global.baseUrl = "example.com";
