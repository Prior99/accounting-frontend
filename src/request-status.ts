export enum RequestStatus {
    PENDING = "pending",
    IN_PROGRESS = "in_progress",
    FAIL = "fail",
    SUCCESS = "success",
}
