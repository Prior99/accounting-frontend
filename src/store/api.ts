import { store } from ".";

declare var baseUrl: string;

type HTTPMethod = "DELETE" | "GET" | "PUT" | "POST";

export async function api(url: string, body?: any, method: HTTPMethod = "GET"): Promise<any> {
    const headers = new Headers();
    if (store.login.loggedIn) {
        headers.append("authorization", store.login.authToken);
    }
    headers.append("content-type", "application/json");
    const fullUrl = `${window.location.protocol}//${baseUrl}${url}`;
    try {
        const response = await fetch(fullUrl, {
            method,
            headers,
            body: JSON.stringify(body),
        });
        const jsonResponse = (await response.json());
        const okay = response.status >= 200 && response.status <= 299;
        if (!okay) {
            store.error.apiError(`Fetching from url "${fullUrl}" resulted in error: "${jsonResponse.message}"`);
            return { okay };
        }
        return {
            okay,
            data: jsonResponse.data,
        };
    } catch (err) {
        store.error.apiError(`Failed to fetch from url "${fullUrl}".`, err);
        return { okay: false };
    }
}
