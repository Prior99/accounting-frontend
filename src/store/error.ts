import { observable, computed, action } from "mobx";

export interface DisplayableError {
    dismissed: boolean;
    message: string;
    origin: "api" | "internal";
    error?: Error;
}

export class ErrorStore {
    constructor() {
        this.apiError = this.apiError.bind(this);
        this.internalError = this.internalError.bind(this);
        this.dismissCurrent = this.dismissCurrent.bind(this);
    }

    @observable public errors: DisplayableError[] = [];

    @computed public get current() {
        return this.errors.find(error => !error.dismissed);
    }

    @action
    public dismissCurrent() {
        if (typeof this.current === "undefined") {
            return;
        }
        this.current.dismissed = true;
    }

    @action
    public apiError(message: string, error?: Error) {
        this.errors.push({ dismissed: false, message, origin: "api", error });
    }

    @action
    public internalError(message: string, error?: Error) {
        this.errors.push({ dismissed: false, message, origin: "internal", error });
    }
}
