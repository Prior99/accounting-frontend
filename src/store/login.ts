import { observable, computed, action } from "mobx";
import { api } from "./api";
import { bind } from "bind-decorator";
import { browserHistory } from "browser-history";
import { routeDashboard } from "routing";
import { RequestStatus } from "request-status";

const softwareVersion = 1;
const localStorageIdentifier = "accounting-software-login";

interface LocalStorageLogin {
    storageVersion: number;
    date: string;
    authToken: string;
}

export class LoginStore {
    @observable public authToken: string;
    @observable public status = RequestStatus.PENDING;

    constructor() {
        this.load();
    }

    @computed
    public get loggedIn() {
        return typeof this.authToken !== "undefined";
    }

    @bind
    @action
    public async onLogin(email: string, password: string) {
        this.status = RequestStatus.IN_PROGRESS;
        const { data, okay } = await api("/auth/login", { email, password }, "POST", true);
        if (okay && data.token) {
            this.authToken = data.token;
            this.save();
            browserHistory.replace(routeDashboard());
            this.status = RequestStatus.SUCCESS;
        } else {
            this.status = RequestStatus.FAIL;
        }
    }

    @bind
    public save() {
        const deserialized: LocalStorageLogin = {
            storageVersion: softwareVersion,
            date: new Date().toString(),
            authToken: this.authToken,
        };
        const serialized = JSON.stringify(deserialized);
        localStorage.setItem(localStorageIdentifier, serialized);
    }

    @bind
    public clearStorage() {
        localStorage.removeItem(localStorageIdentifier);
    }

    @bind
    public load() {
        const serialized = localStorage.getItem(localStorageIdentifier);
        if (serialized === null) { // tslint:disable-line
            return;
        }
        let deserialized: LocalStorageLogin;
        try {
            deserialized = JSON.parse(serialized);
        } catch (err) {
            this.clearStorage();
            return;
        }
        if (deserialized.storageVersion !== softwareVersion) {
            this.clearStorage();
            return;
        }
        this.authToken = deserialized.authToken;
        browserHistory.replace(routeDashboard());
    }
}
