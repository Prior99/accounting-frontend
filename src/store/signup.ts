import { observable, computed, action } from "mobx";
import { api } from "./api";
import { bind } from "bind-decorator";
import { store } from ".";
import { RequestStatus } from "request-status";

export class SignupStore {
    @observable public status = RequestStatus.PENDING;

    @bind
    @action
    public async onSignup(email: string, password: string) {
        this.status = RequestStatus.IN_PROGRESS;
        const result = await api("/auth/register", { email, password }, "POST", true);
        const { data } = result;
        if (result.okay) {
            await store.login.onLogin(email, password);
            this.status = store.login.status;
        } else {
            this.status = RequestStatus.FAIL;
        }
    }
}
