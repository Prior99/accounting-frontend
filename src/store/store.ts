import { ApiStore } from "./api";
import { History } from "history";

export class Store {
    protected api: ApiStore;
    protected browserHistory: History;

    constructor(api: ApiStore, browserHistory: History) {
        this.api = api;
        this.browserHistory = browserHistory;
    }
}
