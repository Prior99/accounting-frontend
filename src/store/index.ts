import { LoginStore } from "./login";
import { SignupStore } from "./signup";
import { ErrorStore } from "./error";
export * from "./error";

export interface Store {
    login: LoginStore;
    error: ErrorStore;
    signup: SignupStore;
}

export let store: Store;

export function createStore(): Store {
    return {
        login: new LoginStore(),
        error: new ErrorStore(),
        signup: new SignupStore()
    };
}

export function resetStore() {
    store = createStore();
}

resetStore();

if (typeof window !== "undefined") {
    (window as any).store = store;
    (window as any).createStore = createStore;
    (window as any).resetStore = resetStore;
}
