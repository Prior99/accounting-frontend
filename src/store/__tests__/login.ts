import "mocks/base-url";

import "isomorphic-fetch";
import { store, resetStore } from "store";
import * as fetchMock from "fetch-mock";
import { observable } from "mobx";
import * as HTTP from "http-status-codes";
import { RequestStatus } from "request-status";
import { browserHistory } from "browser-history";

afterEach(() => {
    fetchMock.restore();
    localStorage.clear();
    resetStore();
});

const someEmail = "user@example.com";
const somePassword = "somepassword";
const someAuthToken = "some-auth-token";

test("The `LoginStore` has `loggedIn` `true` when an `authToken` is present", () => {
    store.login.authToken = someAuthToken;
    expect(store.login.loggedIn).toBe(true);
});

test("The `LoginStore` has `loggedIn` `true` after performing a successfull login", async () => {
    fetchMock.mock(`about://example.com/auth/login`, {
        status: HTTP.OK,
        body: JSON.stringify({
            token: someAuthToken
        }),
    });
    browserHistory.replace = jest.fn();

    expect(store.login.status).toBe(RequestStatus.PENDING);
    await store.login.onLogin(someEmail, somePassword);
    expect(store.login.status).toBe(RequestStatus.SUCCESS);
    expect(store.login.loggedIn).toBe(true);
    expect(store.login.authToken).toBe(someAuthToken);
    // `localStorage` is getting updated.
    expect(localStorage["accounting-software-login"]).toMatchSnapshot();
    // Redirect to `/login`.
    expect(browserHistory.replace).toHaveBeenCalled();
});

test("The `LoginStore` has `status` `FAIL` after performing a failed login", async () => {
    expect(store.login.status).toBe(RequestStatus.PENDING);
    await store.login.onLogin(someEmail, somePassword);
    expect(store.login.status).toBe(RequestStatus.FAIL);
    expect(store.login.loggedIn).toBe(false);
});

test("The `LoginStore` loads the `localStorage`", () => {
    const someDateString = new Date().toString();
    localStorage.setItem("accounting-software-login", JSON.stringify({
        date: someDateString,
        storageVersion: 1,
        authToken: someAuthToken,
    }));
    resetStore();
    expect(store.login.authToken).toBe(someAuthToken);
});

test("The `LoginStore` doesn't load a `localStorage` with an invalid version", () => {
    const someDateString = new Date().toString();
    localStorage.setItem("accounting-software-login", JSON.stringify({
        date: someDateString,
        storageVersion: -90,
        authToken: someAuthToken,
    }));
    resetStore();
    expect(store.login.authToken).toBeUndefined();
    expect(localStorage.getItem("accounting-software-login")).toBeUndefined();
});

test("The `LoginStore` doesn't load a corrupted `localStorage`", () => {
    const someDateString = new Date().toString();
    localStorage.setItem("accounting-software-login", JSON.stringify({
        date: someDateString,
        storageVersion: 1,
        authToken: someAuthToken,
    }).substr(0, 10));
    resetStore();
    expect(store.login.authToken).toBeUndefined();
    expect(localStorage.getItem("accounting-software-login")).toBeUndefined();
});
