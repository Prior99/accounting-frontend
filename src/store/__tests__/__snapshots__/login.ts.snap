// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The \`LoginStore\` has \`loggedIn\` \`true\` after performing a successfull login 1`] = `undefined`;
